mypy/errors.py

```diff
@@ -271,6 +283,36 @@ def yield_uselessness_error_infos(self) -> Iterator[tuple[str, Context, ErrorCod
             context.end_column = error_info[5]
             yield error_info[1], context, error_info[0]
 
+    def yield_nonoverlapping_types(
+        self,
+    ) -> Iterator[tuple[tuple[list[Type], list[Type]], str, Context]]:
+        """Report expressions were non-overlapping types were detected for all iterations
+        were the expression was reachable."""
+
+        selected = set()
+        for candidate in set(chain(*self.nonoverlapping_types)):
```

Comment:
I'd usually use 

```suggestion
        for candidate in set(chain.from_iterable(self.nonoverlapping_types)):
```

but it's up to you, shouldn't matter much here with a small input sequence